src/fabric.c

```diff
@@ -762,6 +791,81 @@ struct fi_info *ofi_allocinfo_internal(void)
 	return NULL;
 }
 
+static int ofi_dup_nic(struct fid_nic *nic, fid_t *target)
+{
+	struct fid_nic *nic_dup;
+
+	nic_dup = calloc(1, sizeof(*nic_dup));
+	if (!nic_dup)
+		return -FI_ENOMEM;
+	nic_dup->fid.fclass = nic->fid.fclass;
+
+	*target = &nic_dup->fid;
+
+	struct fi_device_attr *da, *da_dup;
+	da_dup = calloc(1, sizeof(*da_dup));
+	if (!da_dup) {
+		goto fail;
+	}
+	nic_dup->device_attr = da_dup;
+
+	// Copy the device_attr
+
+#define MEMBER_STRDUP(SRC, TARGET, FIELD)			\
+	do {							\
+		if (SRC->FIELD) {				\
+			TARGET->FIELD = strdup(SRC->FIELD);	\
+			if (!TARGET->FIELD)			\
+				goto fail;			\
+		}						\
+	} while(0)
+
+	da = nic->device_attr;
+	MEMBER_STRDUP(da, da_dup, name);
+	MEMBER_STRDUP(da, da_dup, device_id);
+	MEMBER_STRDUP(da, da_dup, device_version);
+	MEMBER_STRDUP(da, da_dup, vendor_id);
+	MEMBER_STRDUP(da, da_dup, driver);
+	MEMBER_STRDUP(da, da_dup, firmware);
+
+	// Copy the bus_attr
+
+	struct fi_bus_attr *ba_dup;
+	ba_dup = calloc(1, sizeof(*ba_dup));
+	if (!ba_dup)
+		goto fail;
+	nic_dup->bus_attr = ba_dup;
+
+	// Just do a deep copy: fi_bus_attr doesn't contain any
+	// pointers.
+	*ba_dup = *nic->bus_attr;
+
+	// Copy the link_attr
+
+	struct fi_link_attr *la, *la_dup;
+	la_dup = calloc(1, sizeof(*la_dup));
+	if (!la_dup)
+		goto fail;
+	nic_dup->link_attr = la_dup;
+
+	la = nic->link_attr;
+	MEMBER_STRDUP(la, la_dup, address);
+	MEMBER_STRDUP(la, la_dup, network_type);
+	la_dup->mtu = la->mtu;
+	la_dup->speed = la->speed;
+	la_dup->state = la->state;
+
```

Comment:
Not that it's very important, but should you `#undef MEMBER_STRDUP` here?

Reply:
Sure; I have no strong opinions about this.  Done.